include/nbl/video/IGPUQueue.h

```diff
@@ -73,6 +73,26 @@ inline bool IGPUQueue::submit(uint32_t _count, const SSubmitInfo* _submits, IGPU
     return true;
 }
 
+class CThreadSafeGPUQueueAdapter : IGPUQueue
+{
+protected:
+    IGPUQueue* originalQueue;
+    std::mutex m;
+public:
+    CThreadSafeGPUQueueAdapter(IGPUQueue* originalQueue) : IGPUQueue(nullptr, 0, E_CREATE_FLAGS::ECF_PROTECTED_BIT, 0.f), originalQueue(originalQueue) {}
```

Comment:
you need to call the ctor with sane (copied or derived from the underlying queue) arguments

Reply:
take a `core::smart_refctd_ptr<IGPUQueue>&& originalQueue` instead of a "dumb" pointer

Reply:
> you need to call the ctor with sane (copied or derived from the underlying queue) arguments

What for? I thought that this adapter is supposed to only use inheritance for the function overriding purposes. If i call the ctor with the data from the original queues, i'll just have two queues.

Reply:
> If i call the ctor with the data from the original queues, i'll just have two queues.

You're calling the constructor of the base class (IGPUQueue) anyway.

look at the code, see what it does... it doesn't magically create a queue

Reply:
Ah, good point.

Reply:
resolved